File: backend/src/main/java/at/ac/tuwien/sepr/groupphase/backend/service/LocationService.java
package at.ac.tuwien.sepr.groupphase.backend.service;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.location.DetailedLocationDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.location.LocationCreateDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.location.SimpleLocationDto;
import at.ac.tuwien.sepr.groupphase.backend.entity.Location;

import java.util.List;

/**
 * Service for managing locations.
 */
public interface LocationService {

    /**
     * Find all locations.
     *
     * @return list of all locations
     */
    List<Location> findAll();

    /**
     * Find location by ID.
     *
     * @param id the location ID
     * @return the location entity
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException if location not found
     */
    Location findById(Long id);

    /**
     * Create a new location.
     *
     * @param location the location to create
     * @return the created location
     */
    Location create(Location location);

    /**
     * Search locations by multiple criteria (all optional, case-insensitive partial match).
     * Combines all provided criteria with AND logic.
     *
     * @param name    the location name to search for (optional)
     * @param street  the street name to search for (optional)
     * @param city    the city name to search for (optional)
     * @param zipCode the exact zip code to match (optional)
     * @return list of matching locations
     */
    List<Location> searchLocations(String name, String street, String city, Integer zipCode);

    /**
     * Get location by ID with sectors as detailed DTO.
     *
     * @param id the location ID
     * @return the location as detailed DTO including all sectors
     * @throws at.ac.tuwien.sepr.groupphase.backend.exception.NotFoundException if location not found
     */
    DetailedLocationDto findByIdAsDto(Long id);

    /**
     * Create location from DTO.
     *
     * @param dto the location creation data
     * @return the created location as DTO
     */
    SimpleLocationDto createFromDto(LocationCreateDto dto);

    /**
     * Search locations by criteria and return as DTOs.
     * All criteria are optional and combined with AND logic.
     *
     * @param name    the location name (optional)
     * @param street  the street name (optional)
     * @param city    the city name (optional)
     * @param zipCode the zip code (optional)
     * @return list of matching locations as DTOs
     */
    List<SimpleLocationDto> searchLocationsAsDto(String name, String street, String city, Integer zipCode);

    /**
     * Get all locations as DTOs.
     *
     * @return list of all locations as DTOs
     */
    List<SimpleLocationDto> findAllAsDto();
}

File: frontend/src/app/services/location.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Globals} from '../global/globals';
import {Location} from '../dtos/location';
import {Event} from '../dtos/event';

/**
 * Service for managing location/venue-related operations.
 * Provides functionality for location search, retrieval, and querying events by location.
 * Handles communication with the backend locations API endpoint.
 */
@Injectable({
  providedIn: 'root'
})
export class LocationService {
  private baseUri: string = this.globals.backendUri + '/locations';

  constructor(
    private httpClient: HttpClient,
    private globals: Globals
  ) {
  }

  /**
   * Retrieves all locations/venues from the system.
   *
   * @returns Observable of all locations as an array
   */
  getAllLocations(): Observable<Location[]> {
    return this.httpClient.get<Location[]>(this.baseUri);
  }

  /**
   * Searches for locations using multiple filter criteria.
   * All parameters are optional and can be combined for advanced filtering.
   * Search supports partial matching for text fields.
   *
   * @param name - Optional venue name search term (supports partial matching)
   * @param street - Optional street address search term (supports partial matching)
   * @param city - Optional city name search term (supports partial matching)
   * @param zipCode - Optional postal/ZIP code for exact matching
   * @returns Observable of locations matching the search criteria
   */
  searchLocations(name?: string, street?: string, city?: string, zipCode?: number): Observable<Location[]> {
    let params = new HttpParams();

    if (name) params = params.set('name', name);

    if (street) params = params.set('street', street);

    if (city) params = params.set('city', city);

    if (zipCode) params = params.set('zipCode', zipCode.toString());

    return this.httpClient.get<Location[]>(`${this.baseUri}/search`, {params});
  }

  /**
   * Retrieves detailed information for a specific location by its ID.
   * Includes venue details such as name, address, and hall/sector information.
   *
   * @param id - The unique identifier of the location
   * @returns Observable of the location details
   * @throws NotFoundException if the location with the given ID does not exist
   */
  getLocationById(id: number): Observable<Location> {
    return this.httpClient.get<Location>(`${this.baseUri}/${id}`);
  }

  /**
   * Retrieves all events scheduled at a specific location/venue.
   * Useful for displaying venue-specific event listings.
   *
   * @param locationId - The unique identifier of the location
   * @returns Observable of all events at the specified location
   * @throws NotFoundException if the location does not exist
   */
  getEventsByLocation(locationId: number): Observable<Event[]> {
    return this.httpClient.get<Event[]>(`${this.baseUri}/${locationId}/events`);
  }
}
